scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -287,19 +287,34 @@
 (defn progress-aggregator
   "Aggregates the progress state specified in `data` into the current progress state `instance-id->progress-state`.
    It drops messages if the aggregated state has more than `pending-progress-threshold` different entries.
-   It returns the new progress state."
+   It returns the new progress state.
+   The aggregator also makes a local best-effort to avoid updating the progress state of individual instances with
+   stale data based on the value of progress-sequence in the message.
+   We avoid this stale check in the datomic transaction as it is relatively expensive to perform the checks on a per
+   instance basis in the query."
   [pending-progress-threshold instance-id->progress-state {:keys [instance-id] :as data}]
   (meters/mark! progress-aggregator-message-rate)
   (if (or (< (count instance-id->progress-state) pending-progress-threshold)
           (contains? instance-id->progress-state instance-id))
-    (let [progress-state (select-keys data [:progress-message :progress-percent])
-          instance-id->progress-state' (assoc instance-id->progress-state instance-id progress-state)]
-      (let [old-count (count instance-id->progress-state)
-            new-count (count instance-id->progress-state')]
-        (when (zero? old-count)
-          (counters/clear! progress-aggregator-pending-states-count))
-        (counters/inc! progress-aggregator-pending-states-count (- new-count old-count)))
-      instance-id->progress-state')
+    (if (integer? (:progress-sequence data))
+      (let [progress-state (select-keys data [:progress-message :progress-percent :progress-sequence])
+            instance-id->progress-state' (update instance-id->progress-state instance-id
+                                                 (fn [current-state]
+                                                   (let [new-progress-sequence (:progress-sequence progress-state)
+                                                         old-progress-sequence (:progress-sequence current-state)]
+                                                     (if (or (nil? current-state)
+                                                             (< old-progress-sequence new-progress-sequence))
```

Comment:
Unclear how effective this is giving that the latest sequence is being flushed out with every write to Datomic.

Reply:
I wasn't sure how much we want to protect against out-of-order messages. If we want to preserve across write, probably need to introduce an LRU cache to store the task to progress sequence mapping. Do you want me to go down that path?

Reply:
We now preserve the `instance-id->sequence` values i  the `sequence-cache-store` that persist across datomic writes.